File: Platforms/Windows/Reign.Video.D3D9/BlendStateDesc.cpp
#include "pch.h"
#include "BlendState.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Properties
	DWORD BlendStateDesc::RenderTargetWriteMask::get() {return renderTargetWriteMask;}

	bool BlendStateDesc::BlendEnable::get() {return blendEnable;}
	D3DBLENDOP BlendStateDesc::BlendOp::get() {return blendOp;}
	D3DBLEND BlendStateDesc::SrcBlend::get() {return srcBlend;}
	D3DBLEND BlendStateDesc::DestBlend::get() {return destBlend;}

	bool BlendStateDesc::BlendEnableAlpha::get() {return blendEnableAlpha;}
	D3DBLENDOP BlendStateDesc::BlendOpAlpha::get() {return blendOpAlpha;}
	D3DBLEND BlendStateDesc::SrcBlendAlpha::get() {return srcBlendAlpha;}
	D3DBLEND BlendStateDesc::DestBlendAlpha::get() {return destBlendAlpha;}
	#pragma endregion

	#pragma region Constructors
	BlendStateDesc::BlendStateDesc(BlendStateTypes type)
	{
		renderTargetWriteMask = 0xFFFFFFFF;

		switch (type)
		{
			case (BlendStateTypes::None):
				blendEnable = false;
				blendOp = D3DBLENDOP_ADD;
				srcBlend = D3DBLEND_ONE;
				destBlend = D3DBLEND_ZERO;

				blendEnableAlpha = false;
				blendOpAlpha = D3DBLENDOP_ADD;
				srcBlendAlpha = D3DBLEND_ONE;
				destBlendAlpha = D3DBLEND_ZERO;
				break;

			case (BlendStateTypes::Add):
				blendEnable = true;
				blendOp = D3DBLENDOP_ADD;
				srcBlend = D3DBLEND_ONE;
				destBlend = D3DBLEND_ONE;

				blendEnableAlpha = false;
				blendOpAlpha = D3DBLENDOP_ADD;
				srcBlendAlpha = D3DBLEND_ONE;
				destBlendAlpha = D3DBLEND_ONE;
				break;

			case (BlendStateTypes::Subtract):
				blendEnable = true;
				blendOp = D3DBLENDOP_SUBTRACT;
				srcBlend = D3DBLEND_ONE;
				destBlend = D3DBLEND_ONE;

				blendEnableAlpha = false;
				blendOpAlpha = D3DBLENDOP_SUBTRACT;
				srcBlendAlpha = D3DBLEND_ONE;
				destBlendAlpha = D3DBLEND_ONE;
				break;

			case (BlendStateTypes::RevSubtract):
				blendEnable = true;
				blendOp = D3DBLENDOP_REVSUBTRACT;
				srcBlend = D3DBLEND_ONE;
				destBlend = D3DBLEND_ONE;

				blendEnableAlpha = false;
				blendOpAlpha = D3DBLENDOP_REVSUBTRACT;
				srcBlendAlpha = D3DBLEND_ONE;
				destBlendAlpha = D3DBLEND_ONE;
				break;

			case (BlendStateTypes::Alpha):
				blendEnable = true;
				blendOp = D3DBLENDOP_ADD;
				srcBlend = D3DBLEND_SRCALPHA;
				destBlend = D3DBLEND_INVSRCALPHA;

				blendEnableAlpha = false;
				blendOpAlpha = D3DBLENDOP_ADD;
				srcBlendAlpha = D3DBLEND_SRCALPHA;
				destBlendAlpha = D3DBLEND_INVSRCALPHA;
				break;

			default:
				Debug::ThrowError(L"BlendStateDesc", L"Unsuported BlendStateType");
				break;
		}
	}
	#pragma endregion
}
}
}

File: Platforms/Win32/Reign.Video.D3D9/States/BlendStateDesc.cs
using Reign.Video;
using Reign.Core;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class BlendStateDesc : IBlendStateDesc
	{
		#region Properties
		internal BlendStateDescCom com;
		#endregion

		#region Constructors
		public BlendStateDesc(BlendStateTypes type)
		{
			uint renderTargetWriteMask = 0xFFFFFFFF;
			bool blendEnable = false, blendEnableAlpha = false;
			REIGN_D3DBLENDOP blendOp = REIGN_D3DBLENDOP.ADD, blendOpAlpha = REIGN_D3DBLENDOP.ADD;
			REIGN_D3DBLEND srcBlend = REIGN_D3DBLEND.ONE, dstBlend = REIGN_D3DBLEND.ONE, srcBlendAlpha = REIGN_D3DBLEND.ONE, dstBlendAlpha = REIGN_D3DBLEND.ONE;

			switch (type)
			{
				case BlendStateTypes.None:
					blendEnable = false;
					blendOp = REIGN_D3DBLENDOP.ADD;
					srcBlend = REIGN_D3DBLEND.ONE;
					dstBlend = REIGN_D3DBLEND.ZERO;

					blendEnableAlpha = false;
					blendOpAlpha = REIGN_D3DBLENDOP.ADD;
					srcBlendAlpha = REIGN_D3DBLEND.ONE;
					dstBlendAlpha = REIGN_D3DBLEND.ZERO;
					break;

				case BlendStateTypes.Add:
					blendEnable = true;
					blendOp = REIGN_D3DBLENDOP.ADD;
					srcBlend = REIGN_D3DBLEND.ONE;
					dstBlend = REIGN_D3DBLEND.ONE;

					blendEnableAlpha = false;
					blendOpAlpha = REIGN_D3DBLENDOP.ADD;
					srcBlendAlpha = REIGN_D3DBLEND.ONE;
					dstBlendAlpha = REIGN_D3DBLEND.ONE;
					break;

				case BlendStateTypes.Subtract:
					blendEnable = true;
					blendOp = REIGN_D3DBLENDOP.SUBTRACT;
					srcBlend = REIGN_D3DBLEND.ONE;
					dstBlend = REIGN_D3DBLEND.ONE;

					blendEnableAlpha = false;
					blendOpAlpha = REIGN_D3DBLENDOP.SUBTRACT;
					srcBlendAlpha = REIGN_D3DBLEND.ONE;
					dstBlendAlpha = REIGN_D3DBLEND.ONE;
					break;

				case BlendStateTypes.RevSubtract:
					blendEnable = true;
					blendOp = REIGN_D3DBLENDOP.REVSUBTRACT;
					srcBlend = REIGN_D3DBLEND.ONE;
					dstBlend = REIGN_D3DBLEND.ONE;

					blendEnableAlpha = false;
					blendOpAlpha = REIGN_D3DBLENDOP.REVSUBTRACT;
					srcBlendAlpha = REIGN_D3DBLEND.ONE;
					dstBlendAlpha = REIGN_D3DBLEND.ONE;
					break;

				case BlendStateTypes.Alpha:
					blendEnable = true;
					blendOp = REIGN_D3DBLENDOP.ADD;
					srcBlend = REIGN_D3DBLEND.SRCALPHA;
					dstBlend = REIGN_D3DBLEND.INVSRCALPHA;

					blendEnableAlpha = false;
					blendOpAlpha = REIGN_D3DBLENDOP.ADD;
					srcBlendAlpha = REIGN_D3DBLEND.SRCALPHA;
					dstBlendAlpha = REIGN_D3DBLEND.INVSRCALPHA;
					break;

				default:
					Debug.ThrowError("BlendStateDesc", "Unsuported BlendStateType");
					break;
			}

			com = new BlendStateDescCom(renderTargetWriteMask, blendEnable, blendOp, srcBlend, dstBlend, blendEnableAlpha, blendOpAlpha, srcBlendAlpha, dstBlendAlpha);
		}
		#endregion
	}
}
